### PR TITLE
Upgrade langchain

### PR DESCRIPTION
**Pull request by ross-p-smith**
## Purpose
Both `langchain` and `langchain-community` needed upgrading. `langchain-community` needs to use `langchain-openai` and then this causes us to upgrade the version of `openai`.

The new versions of these packages no longer use the environment variable OPENAI_API_BASE, but rather insist on using AZURE_OPENAI_ENDPOINT

I have re-ordered the environment variables in the bicep files so that they are a bit easier to see and find.

Relates to #417 

## Does this introduce a breaking change?
```
[ ] Yes
[x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[x] Other... Please describe:
```

### CODE REVIEW

File: infra/main.bicep
Comment (ross-p-smith) [inline infra/main.bicep:1]: The changes here are just order

File: code/backend/batch/utilities/helpers/EnvHelper.py [hunk 1]
```diff
@@ -83,7 +83,8 @@ def __init__(self, **kwargs) -> None:
         # Otherwise, keys are obtained from environment variables.
         if self.AZURE_AUTH_TYPE == "rbac":
             self.AZURE_SEARCH_KEY = None
-            self.AZURE_OPENAI_KEY = ""
+            self.AZURE_OPENAI_KEY = ""  # TODO: See if this is still used.
```
Comment (adamdougal) [inline code/backend/batch/utilities/helpers/EnvHelper.py:86]: Left over TODO

File: code/backend/batch/utilities/helpers/EnvHelper.py [hunk 2]
```diff
@@ -96,22 +97,28 @@ def __init__(self, **kwargs) -> None:
...
+            f"https://{self.AZURE_OPENAI_RESOURCE}.openai.azure.com/",
         )
+        os.environ["AZURE_OPENAI_ENDPOINT"] = self.AZURE_OPENAI_ENDPOINT
+        os.environ["AZURE_OPENAI_API_KEY"] = self.AZURE_OPENAI_API_KEY
         self.OPENAI_API_TYPE = "azure" if self.AZURE_AUTH_TYPE == "keys" else "azure_ad"
         self.OPENAI_API_KEY = self.AZURE_OPENAI_KEY
         self.OPENAI_API_VERSION = self.AZURE_OPENAI_API_VERSION
         os.environ["OPENAI_API_TYPE"] = self.OPENAI_API_TYPE
-        os.environ["OPENAI_API_BASE"] = self.OPENAI_API_BASE
         os.environ["OPENAI_API_KEY"] = self.OPENAI_API_KEY
```
Comment (adamdougal) [inline code/backend/batch/utilities/helpers/EnvHelper.py:122]: Is this still needed now we're setting `AZURE_OPENAI_API_KEY`?

File: code/tests/utilities/test_LangChainAgent.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
+from unittest.mock import MagicMock
+
+from backend.batch.utilities.orchestrator.LangChainAgent import LangChainAgent
+from backend.batch.utilities.common.Answer import Answer
+
+
+class Testing_LangChainAgent(LangChainAgent):
+    def __init__(self) -> None:
```
Comment (adamdougal) [inline code/tests/utilities/test_LangChainAgent.py:8]: Nice way of overriding values when you can't provide them via the constructor! Do we need to call `super().__init__()` here though? I can see we're doing it in a few places, including the `LangChainAgent`,

File: code/backend/batch/utilities/helpers/EnvHelper.py (hunk 1 above)
Comment (ross-p-smith) [reply code/backend/batch/utilities/helpers/EnvHelper.py:86]: I was scared to remove it! Will do some testing

File: code/tests/utilities/test_LangChainAgent.py (hunk 3 above)
Comment (ross-p-smith) [reply code/tests/utilities/test_LangChainAgent.py:8]: If I call super.__init then I will also create a whole set of connected server objects which makes it hard to test. The idea was to not initialise the base class and just to test these two methods.

File: .gitignore [hunk 4]
```diff
@@ -413,5 +413,6 @@ temp/
 .whl
 .tar.gz
```
Comment (cecheta) [inline .gitignore:414]:
```suggestion
*.whl
*.tar.gz
```

I know it's unrelated, but could you update this?

**Review by cecheta — APPROVED** [review]

**Pull request merged**